src/platform_impl/linux/wayland/seat/dnd/handlers.rs

```diff
@@ -0,0 +1,136 @@
+    path::PathBuf,
+};
+
+use percent_encoding::percent_decode_str;
+use sctk::data_device::{DataOffer, DndEvent};
+use wayland_client::Display;
+
+use crate::{
+    dpi::PhysicalPosition,
+    event::WindowEvent,
+    platform_impl::wayland::{event_loop::WinitState, make_wid, DeviceId},
+};
+
+use super::DndInner;
+
+const MIME_TYPE: &str = "text/uri-list";
+
+pub(super) fn handle_dnd(event: DndEvent<'_>, inner: &mut DndInner, winit_state: &mut WinitState) {
+    match event {
+        DndEvent::Enter {
+            offer: Some(offer),
+            surface,
+            x,
+            y,
+            ..
+        } => {
+            let window_id = make_wid(&surface);
+
+            if let Ok(paths) = parse_offer(&winit_state.display, offer) {
+                if !paths.is_empty() {
+                    offer.accept(Some(MIME_TYPE.into()));
+
+                    winit_state.event_sink.push_window_event(
+                        WindowEvent::CursorEntered {
+                            device_id: crate::event::DeviceId(
+                                crate::platform_impl::DeviceId::Wayland(DeviceId),
+                            ),
+                        },
+                        window_id,
+                    );
+                    winit_state.event_sink.push_window_event(
+                        WindowEvent::CursorMoved {
+                            device_id: crate::event::DeviceId(
+                                crate::platform_impl::DeviceId::Wayland(DeviceId),
+                            ),
+                            position: PhysicalPosition::new(x, y),
+                            modifiers: Default::default(),
+                        },
+                        window_id,
+                    );
+
+                    for path in paths {
+                        winit_state
+                            .event_sink
+                            .push_window_event(WindowEvent::HoveredFile(path), window_id);
+                    }
+                    inner.window_id = Some(window_id);
+                }
+            }
+        }
+        DndEvent::Drop { offer: Some(offer) } => {
+            if let Some(window_id) = inner.window_id {
+                inner.window_id = None;
+
+                if let Ok(paths) = parse_offer(&winit_state.display, offer) {
+                    for path in paths {
+                        winit_state
+                            .event_sink
+                            .push_window_event(WindowEvent::DroppedFile(path), window_id);
+                    }
+                }
+            }
+        }
+        DndEvent::Leave => {
+            if let Some(window_id) = inner.window_id {
+                inner.window_id = None;
+
+                winit_state
+                    .event_sink
+                    .push_window_event(WindowEvent::HoveredFileCancelled, window_id);
+                winit_state.event_sink.push_window_event(
+                    WindowEvent::CursorLeft {
```

Comment:
Also not sure how it's related, I'm pretty sure `wl_pointer` will send the right events anyway?

Reply:
That doesn't seem to happen at least on KDE and GNOME. Only when no file is being dragged over.